# Computing per replay configuration the EER on the evaluation dataset


Using the following end-to-end CNN model (the one we submitted in INTERSPEECH), we compute RC wise EER on the evaluation test set.

> models_After_ICASSP/InterSpeech2018_v2.0/testing_best_model_with3sec_RELU/keep_0.5_0.5_relurun8

The evaluation dataset has total of 57 replay configurations. Please refer to the following odyssey paper to see more details on this.

> ***ASVspoof 2017 version 2.0: meta-data analysis and baseline enhancements***

In [4]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [8]:
# The model prediction scores on eval data (genuineFirstSpoof)

model='/homes/bc305/myphd/stage2/deeplearning.experiment1/CNN3//models_After_ICASSP/InterSpeech2018_v2.0/testing_best_model_with3sec_RELU/'
evalScore=model+'/keep_0.5_0.5_relurun8/predictions_original/eval_prediction_new.txt'

In [9]:
# GenuineFirstSpoof protocal we have created is to be used for extracting relevant scores

evProtocal='/homes/bc305/myphd/datasets/ASVSpoof2017_v2.0/protocol_V2/ASVspoof2017_V2_eval.trl_genFirstSpoof.txt'

In [10]:
from eer import find_eers
from python_codes import analyse_conf

In [14]:
scoreFile = evalScore
evalScores = analyse_conf.get_scores(scoreFile)

In [15]:
print(len(evalScores))

13306


### 1)  RC01: ['E18 P18 R12']

In [20]:
confKey = ['E18 P18 R12']
gc=1298
saveFolder='rc01'

# Concatenate all spoof scores
scores_spf =  analyse_conf.get_config_specific_scores(scoreFile, evProtocal, confKey)
scores_gen = evalScores[0:gc]

# Put genuine scores on top
scores = np.hstack((scores_gen, scores_spf))

# Make the labels
labels = np.hstack((np.ones(len(scores_gen)), np.zeros(len(scores_spf))))

print('spoof len: ',len(scores_spf))
scoreSavePath = model+'/keep_0.5_0.5_relurun8/replay_configuration_wise/'


# Save the score file and label file
analyse_conf.save_scores_labels(scoreSavePath, scores, labels, saveFolder)

spoof len:  55


In [21]:
%%bash

python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc01


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   29.8030



### 2)  RC02: ['E08 P20 R19']

In [22]:
confKey = ['E08 P20 R19']
gc=1298
saveFolder='rc02'

# Concatenate all spoof scores
scores_spf =  analyse_conf.get_config_specific_scores(scoreFile, evProtocal, confKey)
scores_gen = evalScores[0:gc]

# Put genuine scores on top
scores = np.hstack((scores_gen, scores_spf))

# Make the labels
labels = np.hstack((np.ones(len(scores_gen)), np.zeros(len(scores_spf))))

print('spoof len: ',len(scores_spf))
scoreSavePath = model+'/keep_0.5_0.5_relurun8/replay_configuration_wise/'


# Save the score file and label file
analyse_conf.save_scores_labels(scoreSavePath, scores, labels, saveFolder)

spoof len:  67


In [23]:
%%bash

python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc02


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   24.8477



### 2)  RC03: ['E08 P21 R25']

In [24]:
confKey = ['E08 P21 R25']
gc=1298
saveFolder='rc03'

# Concatenate all spoof scores
scores_spf =  analyse_conf.get_config_specific_scores(scoreFile, evProtocal, confKey)
scores_gen = evalScores[0:gc]

# Put genuine scores on top
scores = np.hstack((scores_gen, scores_spf))

# Make the labels
labels = np.hstack((np.ones(len(scores_gen)), np.zeros(len(scores_spf))))

print('spoof len: ',len(scores_spf))
scoreSavePath = model+'/keep_0.5_0.5_relurun8/replay_configuration_wise/'


# Save the score file and label file
analyse_conf.save_scores_labels(scoreSavePath, scores, labels, saveFolder)

spoof len:  122


In [25]:
%%bash

python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc03


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   40.1907



In [30]:
# Lets just put all the RC details into a list
# These are in accordance to the RC details given in the baseline paper. Index 0 correspond to RC01

rcList1 = ['E18 P18 R12','E08 P20 R19','E08 P21 R25','E08 P20 R25','E08 P20 R14','E08 P21 R14','E08 P21 R18',
          'E02 P21 R25','E07 P20 R25','E02 P20 R25','E07 P20 R19','E02 P20 R19','E03 P08 R20','E07 P21 R18',
          'E02 P21 R18','E02 P21 R14','E07 P21 R25','E07 P20 R14','E02 P20 R14']

In [31]:
rcList2 = ['E07 P21 R14','E18 P11 R12','E12 P16 R11','E20 P10 R15','E17 P12 R17','E09 P18 R12','E17 P12 R10',
           'E06 P09 R07','E21 P03 R01','E18 P05 R03','E15 P19 R20','E16 P07 R07','E17 P12 R09','E13 P14 R04',
           'E17 P12 R04','E13 P14 R17','E22 P04 R17','E15 P19 R04','E12 P17 R11']

In [36]:
rcList3 = ['E22 P03 R04','E16 P07 R06','E13 P14 R09','E11 P26 R16','E06 P09 R06','E13 P14 R10','E10 P25 R16',
           'E18 P22 R21','E23 P15 R13','E19 P22 R22','E14 P03 R17','E01 P15 R13','E16 P07 R05','E10 P26 R16',
           'E14 P03 R04','E06 P09 R05','E26 P24 R24','E25 P13 R08','E24 P23 R23']

In [43]:
allRC = np.hstack((rcList1,rcList2,rcList3))

In [44]:
print(len(allRC))
print(len(rcList1))

57
19


# Produce the score file and label file for all replay configurations

In [60]:
# Produce the score file and label file for all replay configurations

for i in range(0, len(allRC)):    
    saveFolder = 'rc'+str(i+1)    
        
    confKey = list()
    confKey.append(allRC[i])
        
    gc=1298
    
    # Concatenate all spoof scores
    scores_spf =  analyse_conf.get_config_specific_scores(scoreFile, evProtocal, confKey)
    scores_gen = evalScores[0:gc]

    # Put genuine scores on top
    scores = np.hstack((scores_gen, scores_spf))

    # Make the labels
    labels = np.hstack((np.ones(len(scores_gen)), np.zeros(len(scores_spf))))

    #print('spoof len: ',len(scores_spf))
    spoofLen = len(scores_spf)
    scoreSavePath = model+'/keep_0.5_0.5_relurun8/replay_configuration_wise/'


    # Save the score file and label file
    analyse_conf.save_scores_labels(scoreSavePath, scores, labels, saveFolder+'_'+str(spoofLen))
            

In [61]:
%%bash

# RC04

python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc4_102


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   41.8653



In [66]:
%%bash

# RC05

#ls keep_0.5_0.5_relurun8/replay_configuration_wise/rc5_114/
python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc5_114


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   49.5334



In [68]:
%%bash

# RC06

#ls keep_0.5_0.5_relurun8/replay_configuration_wise/rc6_108/
python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc6_108


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   42.6224



In [69]:
%%bash

# RC07

#ls keep_0.5_0.5_relurun8/replay_configuration_wise/rc7_120/
python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc7_120


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   31.5437



In [70]:
%%bash

# RC08

#ls keep_0.5_0.5_relurun8/replay_configuration_wise/rc8_98/
python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc8_98


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   31.7384



In [71]:
%%bash

# RC09

#ls keep_0.5_0.5_relurun8/replay_configuration_wise/rc9_244/
python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc9_244


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   49.6999



In [85]:
%%bash

# RC10

#ls keep_0.5_0.5_relurun8/replay_configuration_wise/rc10_82/
python eer.py keep_0.5_0.5_relurun8/replay_configuration_wise/rc10_82


                            < M A T L A B (R) >
                  Copyright 1984-2018 The MathWorks, Inc.
                   R2018a (9.4.0.813654) 64-bit (glnxa64)
                             February 23, 2018

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

EER =

   33.1403

Writing to the file !!!


In [102]:
%%bash

# Lets us execute all at once. The EER will be written in the respective directories
python eer.py

Process is terminated.


In [118]:
%%bash
cat /homes/bc305/myphd/stage2/meta-data_analysis/keep_0.5_0.5_relurun8/replay_configuration_wise/rc10_82/eer.txt


--- 		33.14	


In [121]:
def read_all_eers(): 
    b='/homes/bc305/myphd/stage2/meta-data_analysis/keep_0.5_0.5_relurun8/replay_configuration_wise/'
    with open('rc_list.txt') as f:
        print('Result below show rcID_number of spoof files: the EER')
        
        for rc in f:
            file=b+rc.strip()+'/eer.txt'
            with open(file) as f:
                for line in f:
                    eer = line.strip().split(' ')[1]
                    #print(rc.strip()+' : '+ str(eer.strip()))
            
read_all_eers()

Result below show rcID_number of spoof files: the EER


> Result below show **rcID_no of spoof files: the EER**

> rc1_55 : 29.80
 
> rc2_67 : 24.85 

> rc3_122 : 40.19

> rc4_102 : 41.87

> rc5_114 : 49.53

> rc6_108 : 42.62

> rc7_120 : 31.54

> rc8_98 : 31.74

> rc9_244 : 49.70

> rc10_82 : 33.14

> rc11_272 : 41.00

> rc12_75 : 32.71

> rc13_113 : 32.56

> rc14_279 : 42.45

> rc15_150 : 32.47

> rc16_116 : 35.05

> rc17_266 : 46.10

> rc18_265 : 49.96

> rc19_120 : 49.40

> rc20_275 : 47.55

> **rc21_22 : 0.57**

> rc22_183 : 45.22

> rc23_1138 : 37.03

> rc24_179 : 22.05

> rc25_42 : 32.99

> **rc26_184 : 6.13**

> **rc27_96 : 3.43**

>rc28_240 : 29.55

>**rc29_454 : 4.17**

>rc30_74 : 23.01

>**rc31_145 : 3.08**

>rc32_180 : 12.90

>rc33_183 : 10.55

>rc34_181 : 10.23

>rc35_178 : 37.15

>rc36_181 : 20.35

>**rc37_48 : 1.07**

>rc38_184 : 23.62

>rc39_183 : 24.83

>**rc40_116 : 6.07**

>rc41_182 : 23.23

>**rc42_153 : 2.98**

>**rc43_84 : 3.71**

>rc44_179 : 39.84

>rc45_346 : 18.20

>rc46_181 : 17.78

>rc47_342 : 12.00

>**rc48_1200 : 9.07**

>rc49_180 : 41.71

>rc50_748 : 22.10

>**rc51_169 : 8.09**

>rc52_181 : 15.90

>rc53_181 : 29.12

>**rc54_105 : 2.51**

>**rc55_178 : 7.82**

>rc56_182 : 21.25

>rc57_183 : 23.33

In [123]:
# TODO:

# Generate the scp files as per the rc which could be used for analysis later by the SLIME !!

**SUMMARY OF RESULTS **

There are 1298 genuine files added to these spoof files to compute the RC specific EERs

 1. RC01 ['E18 P18 R12']: 55 spoof files, EER = 29.8
 1. RC02: ['E08 P20 R19']: 67 spoof files, EER = 24.87
 1. RC03: ['E08 P21 R25']: 122 spoof files, EER = 40.19
 1. RC04: E08 P20 R25, 102 spoof files, EER = 41.86
 1. RC05: E08 P20 R14, 114 , EER = 49.53
 1. RC06: E08 P21 R14, 108, EER = 42.62
 1. RC07: E08 P21 R18, 120, EER = 31.5
 1. RC08: E02 P21 R25, 98, EER = 31.7
 1. RC09: E07 P20 R25, 244, EER = 49.69
 1. RC10: E02 P20 R25, 82, EER = 33.14


In [124]:
# PRINT ALL THE RC WITH THEIR IDS
for i in range(0, len(allRC)):
    print('RC'+str(i+1)+': ' +allRC[i] )

RC1: E18 P18 R12
RC2: E08 P20 R19
RC3: E08 P21 R25
RC4: E08 P20 R25
RC5: E08 P20 R14
RC6: E08 P21 R14
RC7: E08 P21 R18
RC8: E02 P21 R25
RC9: E07 P20 R25
RC10: E02 P20 R25
RC11: E07 P20 R19
RC12: E02 P20 R19
RC13: E03 P08 R20
RC14: E07 P21 R18
RC15: E02 P21 R18
RC16: E02 P21 R14
RC17: E07 P21 R25
RC18: E07 P20 R14
RC19: E02 P20 R14
RC20: E07 P21 R14
RC21: E18 P11 R12
RC22: E12 P16 R11
RC23: E20 P10 R15
RC24: E17 P12 R17
RC25: E09 P18 R12
RC26: E17 P12 R10
RC27: E06 P09 R07
RC28: E21 P03 R01
RC29: E18 P05 R03
RC30: E15 P19 R20
RC31: E16 P07 R07
RC32: E17 P12 R09
RC33: E13 P14 R04
RC34: E17 P12 R04
RC35: E13 P14 R17
RC36: E22 P04 R17
RC37: E15 P19 R04
RC38: E12 P17 R11
RC39: E22 P03 R04
RC40: E16 P07 R06
RC41: E13 P14 R09
RC42: E11 P26 R16
RC43: E06 P09 R06
RC44: E13 P14 R10
RC45: E10 P25 R16
RC46: E18 P22 R21
RC47: E23 P15 R13
RC48: E19 P22 R22
RC49: E14 P03 R17
RC50: E01 P15 R13
RC51: E16 P07 R05
RC52: E10 P26 R16
RC53: E14 P03 R04
RC54: E06 P09 R05
RC55: E26 P24 R24
RC56: E25 P13 R08
R